File: app/src/main/java/jp/juggler/subwaytooter/view/MaxHeightScrollView.java
package jp.juggler.subwaytooter.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.widget.ScrollView;

import jp.juggler.subwaytooter.R;

public class MaxHeightScrollView extends ScrollView {
	public MaxHeightScrollView( Context context ){
		super( context );
	}
	
	public MaxHeightScrollView( Context context, AttributeSet attrs ){
		super( context, attrs );
		TypedArray a = context.obtainStyledAttributes(attrs ,   R.styleable.MaxHeightScrollView);
		parseAttr(a);
		a.recycle();
		
	}
	
	public MaxHeightScrollView( Context context, AttributeSet attrs, int defStyleAttr ){
		super( context, attrs, defStyleAttr );
		
		TypedArray a = context.obtainStyledAttributes(attrs
			, R.styleable.MaxHeightScrollView, defStyleAttr, 0);
		parseAttr(a);
		a.recycle();
	}
	
	int maxHeight;

	void parseAttr( TypedArray a){
		maxHeight = a.getDimensionPixelSize( R.styleable.MaxHeightScrollView_maxHeight, 0);
	}
	
	@Override protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		if (maxHeight > 0){
			int hSize = MeasureSpec.getSize(heightMeasureSpec);
			int hMode = MeasureSpec.getMode(heightMeasureSpec);
			
			switch (hMode){
			case MeasureSpec.AT_MOST:
				heightMeasureSpec = MeasureSpec.makeMeasureSpec(Math.min(hSize, maxHeight), MeasureSpec.AT_MOST);
				break;
			case MeasureSpec.UNSPECIFIED:
				heightMeasureSpec = MeasureSpec.makeMeasureSpec(maxHeight, MeasureSpec.AT_MOST);
				break;
			case MeasureSpec.EXACTLY:
				heightMeasureSpec = MeasureSpec.makeMeasureSpec(Math.min(hSize, maxHeight), MeasureSpec.EXACTLY);
				break;
			}
		}
		
		super.onMeasure(widthMeasureSpec, heightMeasureSpec);
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/view/MaxHeightScrollView.kt
package jp.juggler.subwaytooter.view

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.widget.ScrollView
import jp.juggler.subwaytooter.R
import kotlin.math.min

class MaxHeightScrollView : ScrollView {
	
	var maxHeight : Int = 0
	
	constructor(context : Context) : super(context)
	constructor(context : Context, attrs : AttributeSet) : super(context, attrs) {
		val a = context.obtainStyledAttributes(attrs, R.styleable.MaxHeightScrollView)
		parseAttr(a)
		a.recycle()
	}
	constructor(context : Context, attrs : AttributeSet, defStyleAttr : Int) : super(context, attrs, defStyleAttr) {
		val a = context.obtainStyledAttributes(attrs, R.styleable.MaxHeightScrollView, defStyleAttr, 0)
		parseAttr(a)
		a.recycle()
	}
	
	private fun parseAttr(a : TypedArray) {
		maxHeight = a.getDimensionPixelSize(R.styleable.MaxHeightScrollView_maxHeight, 0)
	}
	
	override fun onMeasure(widthMeasureSpec : Int, heightMeasureSpec : Int) {
		var heightMeasureSpec2 = heightMeasureSpec
		if(maxHeight > 0) {
			val hSize = MeasureSpec.getSize(heightMeasureSpec)
			
			when(MeasureSpec.getMode(heightMeasureSpec)) {
				MeasureSpec.AT_MOST -> heightMeasureSpec2 = MeasureSpec.makeMeasureSpec(min(hSize, maxHeight), MeasureSpec.AT_MOST)
				MeasureSpec.UNSPECIFIED -> heightMeasureSpec2 = MeasureSpec.makeMeasureSpec(maxHeight, MeasureSpec.AT_MOST)
				MeasureSpec.EXACTLY -> heightMeasureSpec2 = MeasureSpec.makeMeasureSpec(min(hSize, maxHeight), MeasureSpec.EXACTLY)
			}
		}
		
		super.onMeasure(widthMeasureSpec, heightMeasureSpec2)
	}
	
}
